spec/lang/representation.md

```diff
@@ -105,25 +106,38 @@ fn decode_ptr<M: Memory>(bytes: List<AbstractByte<M::Provenance>>) -> Option<Poi
             provenance = None;
         }
     }
-    ret(Pointer { addr, provenance })
+    ret(ThinPointer { addr, provenance })
 }
 
-fn encode_ptr<M: Memory>(ptr: Pointer<M::Provenance>) -> List<AbstractByte<M::Provenance>> {
+fn encode_ptr<M: Memory>(ptr: ThinPointer<M::Provenance>) -> List<AbstractByte<M::Provenance>> {
     let bytes_data = M::T::ENDIANNESS.encode(Unsigned, M::T::PTR_SIZE, ptr.addr).unwrap();
     bytes_data.map(|b| AbstractByte::Init(b, ptr.provenance))
+    
+}
+
+fn encode_ptr_meta<M: Memory>(meta: Option<PointerMeta>) -> List<AbstractByte<M::Provenance>> {
+    match meta {
+        None => list!(),
+        Some(PointerMeta::ElementCount(len)) => {
+            let bytes = M::T::ENDIANNESS.encode(Unsigned, M::T::PTR_SIZE, len).unwrap();
+            bytes.map(|b| AbstractByte::Init(b, None))
+        }
+    }
 }
 
 impl Type {
     fn decode<M: Memory>(Type::Ptr(ptr_type): Self, bytes: List<AbstractByte<M::Provenance>>) -> Option<Value<M>> {
-        let ptr = decode_ptr::<M>(bytes)?;
+        if bytes.len() < M::T::PTR_SIZE.bytes() { throw!(); }
+        let ptr = decode_ptr::<M>(bytes.subslice_with_length(Int::ZERO, M::T::PTR_SIZE.bytes()))?;
         if !ptr_type.addr_valid(ptr.addr) {
             return None;
         }
-        ret(Value::Ptr(ptr))
+        // TODO: Decode wide pointers, but needs slight changes in PtrType & Layout
+        ret(Value::Ptr(ptr.widen(None)))
     }
     fn encode<M: Memory>(Type::Ptr(_): Self, val: Value<M>) -> List<AbstractByte<M::Provenance>> {
         let Value::Ptr(ptr) = val else { panic!() };
-        encode_ptr::<M>(ptr)
+        encode_ptr::<M>(ptr.thin_pointer).iter().chain(encode_ptr_meta::<M>(ptr.metadata).iter()).collect()
```

Comment:
Please let-bind the two components, so that it's not one line doing too many things.